File: services/frontend-web/lib/api/sentiment-accuracy.ts
import apiClient from '../api-client';

/**
 * Types + client for GET /api/v1/settings/ai/sentiment/accuracy
 * (eval-harness-and-card aspect, M5.1 disclosure layer).
 *
 * Mirrors the backend's SentimentAccuracyResponse schema 1:1
 * (src/schemas/sentiment_accuracy.py). This is a disclosure card, not a
 * gated premium feature — kept self-contained per the existing precedent of
 * independent lib/api/*.ts modules (see lib/api/churn-accuracy.ts).
 */

export interface ClassMetrics {
  precision: number;
  recall: number;
  f1: number;
  support: number;
}

export interface ProviderEvalResult {
  provider: string;
  n: number;
  macro_precision: number;
  macro_recall: number;
  macro_f1: number;
  accuracy: number;
  per_class: Record<string, ClassMetrics>;
  confusion_matrix: Record<string, Record<string, number>>;
}

export interface EvalSetResult {
  set_name: string;
  n: number;
  vader: ProviderEvalResult | null;
  transformer: ProviderEvalResult | null;
  macro_f1_delta: number | null;
  meets_target: boolean | null;
}

export interface SentimentAccuracyResponse {
  has_results: boolean;
  generated_at: string | null;
  model_id: string | null;
  model_revision: string | null;
  public: EvalSetResult | null;
  in_domain: EvalSetResult | null;
}

/** Format a 0-1 metric fraction as a whole-number percent string, "—" for null. */
export function formatMetricPercent(n: number | null): string {
  if (n === null) return '—';
  return `${Math.round(n * 100)}%`;
}

/** Format a signed macro-F1 delta (transformer - vader), "—" for null. */
export function formatDelta(delta: number | null): string {
  if (delta === null) return '—';
  const sign = delta >= 0 ? '+' : '-';
  return `${sign}${Math.abs(delta).toFixed(2)}`;
}

/** Fetch the sentiment eval accuracy card data (no plan-gating — disclosure feature). */
export async function getSentimentAccuracy(): Promise<SentimentAccuracyResponse> {
  const response = await apiClient.get<SentimentAccuracyResponse>(
    '/api/v1/settings/ai/sentiment/accuracy'
  );
  return response.data;
}

File: services/backend-api/src/schemas/sentiment_accuracy.py
"""
Pydantic schemas for GET /api/v1/settings/ai/sentiment/accuracy
(eval-harness-and-card aspect, M5.1 disclosure layer).

Mirrors the eval_sentiment.py script's committed JSON artifact
(services/backend-api/eval_results/sentiment_accuracy.json) 1:1 so the
route's parse-or-degrade handler is a thin pass-through.
"""
from datetime import datetime
from typing import Dict, Optional

from pydantic import BaseModel, ConfigDict


class ClassMetrics(BaseModel):
    """Precision/recall/F1/support for a single sentiment class."""

    precision: float
    recall: float
    f1: float
    support: int


class ProviderEvalResult(BaseModel):
    """One provider's (vader or transformer) results on one eval set."""

    provider: str
    n: int
    macro_precision: float
    macro_recall: float
    macro_f1: float
    accuracy: float
    per_class: Dict[str, ClassMetrics]
    confusion_matrix: Dict[str, Dict[str, int]]


class EvalSetResult(BaseModel):
    """Both providers' results on one eval set (public or in_domain), plus the
    transformer-vs-VADER delta and whether it meets the in-domain target."""

    set_name: str
    n: int
    vader: Optional[ProviderEvalResult] = None
    transformer: Optional[ProviderEvalResult] = None
    macro_f1_delta: Optional[float] = None
    meets_target: Optional[bool] = None


class SentimentAccuracyResponse(BaseModel):
    """Response for GET /api/v1/settings/ai/sentiment/accuracy.

    has_results=False (with every other field null/absent) is the honest
    "eval not run yet" state — never a 404/500 (mirrors get_embeddings_status's
    never-raises contract). No organization_id scoping: this is a global,
    offline, reproducible disclosure artifact, not a per-org metric.
    """

    model_config = ConfigDict(protected_namespaces=())

    has_results: bool
    generated_at: Optional[datetime] = None
    model_id: Optional[str] = None
    model_revision: Optional[str] = None
    public: Optional[EvalSetResult] = None
    in_domain: Optional[EvalSetResult] = None
